File: app/src/main/java/com/budgetbuddy/data/repository/DebtRepository.kt
package com.budgetbuddy.data.repository

import android.util.Log
import com.budgetbuddy.data.local.dao.DebtDao
import com.budgetbuddy.data.local.entities.DebtEntity
import com.budgetbuddy.data.local.entities.PayoffStrategy
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "DebtRepo"

data class DebtPayoffMonth(
    val month: Int,
    val debtName: String,
    val payment: Double,
    val remainingBalance: Double
)

@Singleton
class DebtRepository @Inject constructor(
    private val debtDao: DebtDao,
    private val firestoreRepository: FirestoreRepository
) {

    fun getActiveDebts(userId: String): Flow<List<DebtEntity>> = debtDao.getActiveDebts(userId)
    fun getAllDebts(userId: String): Flow<List<DebtEntity>> = debtDao.getAllDebts(userId)

    suspend fun insertDebt(debt: DebtEntity): Long {
        val id = debtDao.insertDebt(debt)
        try { firestoreRepository.saveDebt(debt.userId, debt.copy(id = id)) }
        catch (e: Exception) { Log.w(TAG, "saveDebt failed: ${e.message}") }
        return id
    }

    suspend fun updateDebt(debt: DebtEntity) {
        debtDao.updateDebt(debt)
        try { firestoreRepository.saveDebt(debt.userId, debt) }
        catch (e: Exception) { Log.w(TAG, "updateDebt failed: ${e.message}") }
    }

    suspend fun deleteDebt(debt: DebtEntity) {
        debtDao.deleteDebt(debt)
        try { firestoreRepository.deleteDebt(debt.userId, debt.id) }
        catch (e: Exception) { Log.w(TAG, "deleteDebt failed: ${e.message}") }
    }

    suspend fun markDebtPaidOff(id: Long) {
        debtDao.markDebtPaidOff(id)
        try { debtDao.getDebtById(id)?.let { firestoreRepository.saveDebt(it.userId, it.copy(isPaidOff = true)) } }
        catch (e: Exception) { Log.w(TAG, "markDebtPaidOff sync failed: ${e.message}") }
    }

    /**
     * Compute payoff schedule for a list of debts using snowball or avalanche strategy.
     * Returns a flat list of payments per month (simplified schedule).
     */
    fun computePayoffSchedule(debts: List<DebtEntity>, strategy: PayoffStrategy, extraPayment: Double = 0.0): List<DebtPayoffMonth> {
        if (debts.isEmpty()) return emptyList()

        val sorted = when (strategy) {
            PayoffStrategy.SNOWBALL  -> debts.sortedBy { it.balance }
            PayoffStrategy.AVALANCHE -> debts.sortedByDescending { it.interestRate }
        }

        val balances = sorted.map { it.balance }.toMutableList()
        val schedule = mutableListOf<DebtPayoffMonth>()
        var month = 1

        while (balances.any { it > 0 } && month <= 360) {
            // Freed minimums from fully-paid debts roll over to the current target (the snowball/avalanche cascade)
            val freedMinimums = sorted.indices.filter { balances[it] <= 0 }.sumOf { sorted[it].minimumPayment }
            val targetIndex = balances.indexOfFirst { it > 0 }
            val targetBonus = extraPayment + freedMinimums

            sorted.forEachIndexed { index, debt ->
                if (balances[index] <= 0) return@forEachIndexed
                val monthlyRate = debt.interestRate / 100.0 / 12.0
                val interest = balances[index] * monthlyRate
                val payment = if (index == targetIndex) debt.minimumPayment + targetBonus else debt.minimumPayment
                val actualPayment = minOf(payment, balances[index] + interest)
                balances[index] = maxOf(0.0, balances[index] + interest - actualPayment)
                schedule.add(DebtPayoffMonth(month, debt.name, actualPayment, balances[index]))
            }
            month++
        }
        return schedule
    }
}

File: ios/BudgetBuddy/Repositories/DebtRepository.swift
import Foundation
import SwiftData

final class DebtRepository {
    private let context: ModelContext
    private let firestore: FirestoreRepository

    init(context: ModelContext, firestore: FirestoreRepository) {
        self.context   = context
        self.firestore = firestore
    }

    func all(userId: String) throws -> [BBDebt] {
        let pred = #Predicate<BBDebt> { $0.userId == userId }
        return try context.fetch(FetchDescriptor<BBDebt>(predicate: pred,
                                                         sortBy: [SortDescriptor(\.createdAt, order: .reverse)]))
    }

    func active(userId: String) throws -> [BBDebt] {
        let pred = #Predicate<BBDebt> { $0.userId == userId && !$0.isPaidOff }
        return try context.fetch(FetchDescriptor<BBDebt>(predicate: pred))
    }

    func byId(_ id: UUID) throws -> BBDebt? {
        let pred = #Predicate<BBDebt> { $0.id == id }
        return try context.fetch(FetchDescriptor<BBDebt>(predicate: pred)).first
    }

    func paidOffCount(userId: String) throws -> Int {
        let pred = #Predicate<BBDebt> { $0.userId == userId && $0.isPaidOff }
        return try context.fetchCount(FetchDescriptor<BBDebt>(predicate: pred))
    }

    @discardableResult
    func insert(_ debt: BBDebt) async throws -> BBDebt {
        context.insert(debt)
        try context.save()
        Task { await firestore.saveDebt(userId: debt.userId, debt: debt) }
        return debt
    }

    func update(_ debt: BBDebt) async throws {
        try context.save()
        Task { await firestore.saveDebt(userId: debt.userId, debt: debt) }
    }

    func markPaidOff(_ debt: BBDebt) async throws {
        debt.isPaidOff = true
        debt.balance   = 0
        try context.save()
        Task { await firestore.saveDebt(userId: debt.userId, debt: debt) }
    }

    func delete(_ debt: BBDebt) async throws {
        let userId = debt.userId
        let id     = debt.id
        context.delete(debt)
        try context.save()
        Task { await firestore.deleteDebt(userId: userId, id: id) }
    }

    // MARK: - Payoff schedule (mirrors Android DebtRepository.computePayoffSchedule)

    func computePayoffSchedule(debts: [BBDebt], strategy: PayoffStrategy,
                                extraPayment: Double = 0) -> [DebtPayoffMonth] {
        guard !debts.isEmpty else { return [] }

        let sorted = strategy == .snowball
            ? debts.sorted { $0.balance < $1.balance }
            : debts.sorted { $0.interestRate > $1.interestRate }

        var balances = sorted.map { $0.balance }
        var schedule = [DebtPayoffMonth]()
        var month    = 1

        while balances.contains(where: { $0 > 0 }), month <= 360 {
            let freedMinimums   = sorted.indices.filter { balances[$0] <= 0 }.reduce(0.0) { $0 + sorted[$1].minimumPayment }
            let targetIndex     = balances.firstIndex(where: { $0 > 0 }) ?? 0
            let targetBonus     = extraPayment + freedMinimums

            for (i, debt) in sorted.enumerated() {
                guard balances[i] > 0 else { continue }
                let monthlyRate   = debt.interestRate / 100.0 / 12.0
                let interest      = balances[i] * monthlyRate
                let payment       = i == targetIndex ? debt.minimumPayment + targetBonus : debt.minimumPayment
                let actual        = min(payment, balances[i] + interest)
                balances[i]       = max(0, balances[i] + interest - actual)
                schedule.append(DebtPayoffMonth(month: month, debtName: debt.name,
                                                payment: actual, remainingBalance: balances[i]))
            }
            month += 1
        }
        return schedule
    }
}
